3p/recaptcha.js

```diff
@@ -124,28 +137,63 @@ function initializeIframeMessagingClient(window, grecaptcha, dataObject) {
  * action {string} - action to be dispatched with grecaptcha.execute
  * id {number} - id given to us by a counter in the recaptcha service
  *
+ * @param {Window} window
  * @param {*} grecaptcha
  * @param {Object} data
  */
-function actionTypeHandler(grecaptcha, data) {
-  // TODO: @torch2424: Verify message origin
-  // TODO: @torch2424: Verify sitekey is the same as original
-  const executePromise = grecaptcha.execute(data.sitekey, {
-    action: data.action,
-  });
+function actionTypeHandler(window, grecaptcha, data) {
+  doesOriginDomainMatchIframeSrc(window, data).then(() => {
+    // TODO: @torch2424: Verify sitekey is the same as original
+    const executePromise = grecaptcha.execute(data.sitekey, {
+      action: data.action,
+    });
 
-  // .then() promise pollyfilled by recaptcha api script
-  executePromise./*OK*/then(function(token) {
-    iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-token', dict({
-      'id': data.id,
-      'token': token,
-    }));
-  }, function(err) {
-    user().error(TAG, '%s', err.message);
-    iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-error', dict({
-      'id': data.id,
-      'error': err.message,
-    }));
+    // .then() promise pollyfilled by recaptcha api script
+    executePromise./*OK*/then(function(token) {
+      iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-token', dict({
+        'id': data.id,
+        'token': token,
+      }));
+    }, function(err) {
+      user().error(TAG, '%s', err.message);
+      iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-error', dict({
+        'id': data.id,
+        'error': err.message,
+      }));
+    });
+  }).catch(error => {
+    dev().error(TAG, '%s', error.message);
   });
 }
 
+/**
+ * Function to verify our origin domain from the
+ * parent window.
+ * @param {Window} window
+ * @param {Object} data
+ * @return {!Promise}
+ */
+function doesOriginDomainMatchIframeSrc(window, data) {
+  if (mode.localDev || mode.test) {
+    return Promise.resolve();
+  }
+
+  if (!data.origin) {
+    return Promise.reject(
+        new Error('Could not retreive the origin domain')
+    );
+  }
+
+  return ampToolboxCacheUrl.createCurlsSubdomain(data.origin)
+      .then(curlsSubdomain => {
+        const iframeSrcCurlsSubdomain = window.location.hostname.split('.')[0];
+        if (curlsSubdomain === iframeSrcCurlsSubdomain) {
+          return Promise.resolve();
+        }
+
+        return Promise.reject(
+            new Error('Origin domain does not match Iframe src')
```

Comment:
worth to print out the URLs in error message.

Reply:
Done.